seq2seq/models/attention.py

```diff
@@ -180,3 +206,89 @@ def forward(self, decoder_states, encoder_states):
         attn = out.view(batch_size, dec_seqlen, enc_seqlen)
 
         return attn
+
+class HardGuidance(nn.Module):
+    """
+    Attention method / attentive guidance method for data sets that are annotated with attentive guidance.
+    """
+
+    def forward(self, decoder_states, encoder_states, step, input_lengths, provided_attention):
+        """
+        Forward method that receives provided attentive guidance indices and returns proper
+        attention scores vectors.
+
+        Args:
+            decoder_states (torch.autograd.Variable): Hidden layer of all decoder states (batch, dec_seqlen, hl_size)
+            encoder_states (torch.autograd.Variable): Output layer of all encoder states (batch, dec_seqlen, hl_size)
+            step (int): The current decoder step for unrolled RNN. Set to -1 for rolled RNN
+            input_lengths (list(int)): The input length for all elements in the batch
+            provided_attention (torch.autograd.Variable): Variable containing the provided attentive guidance indices (batch, max_provided_attention_length)
+
+        Returns:
+            torch.autograd.Variable: Attention score vectors (batch, dec_seqlen, hl_size)
+        """
+
+        # decoder_states --> (batch, dec_seqlen, hl_size)
+        # encoder_states --> (batch, enc_seqlen, hl_size)
+        batch_size, enc_seqlen, _ = encoder_states.size()
+        _,          dec_seqlen, _ = decoder_states.size()
+
+        input_lengths = torch.LongTensor(input_lengths)
+        provided_attention = provided_attention.data
+
+        last_encoder_state_indices = input_lengths - 1
+
+        max_input_length = max(input_lengths)
+        max_provided_attention_length = provided_attention.size(1)
+
+        # In the case of rolled RNN
+        if step == -1:
+            attention_indices = provided_attention
+
+            # If there are more decoder states than attentions provided, we extend the provided attentions.
```

Comment:
I am not sure which case this addresses. At training time this shouldn't occur, unless the user made a mistake with the attention targets. This I think would usually concern <eos>, but I think that if <eos> is added at preprocessing time to the input sequence, the user should also provide attention for it explicitly (or otherwise it should be added earlier, when the tabular fields are formed). At testing time, this is similar. Then the only case where I can see that having longer decoder outputs than attention patterns is at free inference, because then there is no explicit stop to the decoder. I think at this point this doesn't warrant all these checks though, because there is not really any hypothesis about what should happen with this in these cases.

Reply:
Your're right, originally i implemented something in the old `HardCoded` method for lookup tables, for the case of free inference. 
For this new method, we don't need this check anymore. It doesn't work anyway in free inference mode